feat: make `backport.py` more flexible for complex pull requests

cfm opened the pull request:
## Status

Ready for review

## Description of Changes

The much-loved `backport.py` script makes 2.5 assumptions that do not hold for complex pull requests like #7143, which we fix here:
1. There are fewer than 30 commits (the default number of results returned by `gh api`) in the pull request.
2. All commits can be merged naïvely via a mass `git cherry-pick -x` (i.e., there are no merge commits requiring `git [cherry-pick → merge] -m`).
    1. *All* of the commits in the pull request should be backported.

## Testing

Depending on the state of your local `release/2.10.1` branch, you can replicate #7257 with:

```sh-session
$ git branch -D release/2.10.1
$ git checkout -b release/2.10.1 release/2.10.0
$ utils/backport.py 7143 2.10.1 origin --skip 6e3d7bed17537fa8f69f941f21a8eec34e8da113
```


## Deployment

Development-only; no deployment considerations.

cfm committed:
feat: "backport.py --skip X" omits commit X from cherry-picking

For example, Weblate-side merge commits from "develop" require "git
cherry-pick -m" and therefore can't be merged en masse even if we wanted
their contents.

cfm committed:
fix: call "gh api --paginate" to fetch all commits in a pull request

Otherwise only the first 30 commits are returned.

legoktm commented on utils/backport.py:
```diff
@@ -16,23 +16,32 @@
+# NB. It's tempting to do something like `set(commit_hashes) -= set(args.skip)`
+# here, but we must preserve the order of commits, and it's not worth pulling
+# in an ordered-set implementation just for this.
+commit_hashes = [commit["sha"] for commit in commits]
```
if you like list comprehensions... :)

```suggestion
commit_hashes = [commit["sha"] for commit in commits if commit["sha"] not in args.skip]
```

cfm replied:
Delightful.  Thanks for the reminder.

legoktm commented on utils/backport.py:
```diff
@@ -16,23 +16,32 @@
+subprocess.check_call(["git", "cherry-pick", "-x"] + commit_hashes)
 if input("OK to push and create PR? [y/N]").lower() != "y":
     sys.exit()
 subprocess.check_call(["git", "push", "-u", remote, branch])
```
In the message body below, let's mention the skipped commits? E.g. `Only contains changes from #{args.pr}, except {commits}.`?

cfm replied:
Good idea.  Added in 234d5bfd7d00a6a868169c3d0fc902a97bce31a3, but do let me know if you see a nicer way to build up the conditional `skip_message` without fancier templating.

legoktm reviewed (changes requested):
Thanks, it LGTM; the one change I'm asking for is if can we mention the skipped commits in the PR body.

cfm committed:
refactor: use list comprehension with inline conditional

Co-authored-by: Kunal Mehta <legoktm@debian.org>

cfm committed:
feat: list skipped commits in pull request

thread: https://github.com/freedomofpress/securedrop/pull/7260#discussion_r1805231906

cfm commented:
I've bumped the `admin-test` suite, which was failing as usual during the v2.10 release process.  With that passing, this should be ready to go, @legoktm.

cfm commented:
Currently blocked on #7339.

cfm commented:
Rebased from `develop` after #7339.

legoktm reviewed (approved):
Thanks!
